feat: add ability to switch and assert themes on iOS and Android: setDarkMode, toggleDarkMode, assertDarkMode & assertLightMode

markrickert opened the pull request:
## Proposed changes

This is a new feature that adds the ability to set the `light`/`dark` mode on iOS and Android platforms.

It adds the following new functionality:

- `setDarkMode` - possible values: `enabled` / `disabled`
- `toggleDarkMode` - swaps between light/dark mode
- `assertDarkMode` - fails the flow if dark mode is not currently enabled
- `assertLightMode` - fails the flow if dark mode is currently enabled

It achieves this by running commands for iOS/Android:

|   | iOS | Android |
|--------|--------|--------|
| `setDarkMode: enabled` | `XCUIDevice.shared.appearance = .dark` (via a new `setAppearance` endpoint on the XCUITest runner) | `adb shell 'cmd uimode night yes'` |
| `setDarkMode: disabled` | `XCUIDevice.shared.appearance = .light` (via the same endpoint) | `adb shell 'cmd uimode night no'` | 

iOS requires iOS 15+ (`XCUIDevice.appearance` isn't available below that); the endpoint returns a precondition error on older runtimes.

Web (`CdpWebDriver`) is a no-op for now — `isDarkModeEnabled`/`setDarkMode` are stubbed out since there's no dark-mode signal to drive on a headless/CDP-controlled browser yet.

## Testing

`./gradlew :maestro-test:test && ./gradlew test` passes 🟢 

I then "ignited" an app that [has a default maestro test in it called `FavoritePodcast.yaml`](https://github.com/infinitered/ignite/blob/master/boilerplate/.maestro/flows/FavoritePodcast.yaml) and replaced the contents with the following, adding calls to `setDarkMode`, `toggleDarkMode`, `assertDarkMode`, and `assertLightMode`

```yaml
# flow: run the login flow and then navigate to the demo podcast list screen, favorite a podcast, and then switch the list to only be favorites.

appId: ${MAESTRO_APP_ID}
env:
  FAVORITES_TEXT: "Switch on to only show favorites" # en.demoPodcastListScreen.accessibility.switch
onFlowStart:
  - runFlow: ../shared/_OnFlowStart.yaml
---
- runFlow: ../shared/_Login.yaml
- setDarkMode: disabled
- assertLightMode
- toggleDarkMode
- toggleDarkMode
- assertLightMode
- tapOn: "Podcast"
- assertVisible: "React Native Radio episodes"
- tapOn:
    text: ${FAVORITES_TEXT}
- setDarkMode: disabled
- assertVisible: "This looks a bit empty"
- tapOn:
    text: ${FAVORITES_TEXT}
    # https://maestro.mobile.dev/troubleshooting/known-issues#android-accidental-double-tap
    retryTapIfNoChange: false
- toggleDarkMode
- assertDarkMode
- repeat:
    times: 2
    commands:
      - scroll
- copyTextFrom:
    text: "RNR .*" # assumes all podcast titles start with RNR
    index: 2 # grab the third one, others might not be fully visible
- toggleDarkMode
- longPressOn: ${maestro.copiedText}
- scrollUntilVisible:
    element:
      text: ${FAVORITES_TEXT}
    direction: UP
    timeout: 50000
    speed: 90
    visibilityPercentage: 100
- tapOn:
    text: ${FAVORITES_TEXT}
- assertVisible: ${maestro.copiedText}
```

Then to test this, i ran the associated application on the iOS simulator and from this projects' directory i ran:

```bash
./maestro record -e MAESTRO_APP_ID=com.foo ../Foo/.maestro/flows/FavoritePodcast.yaml
```

Here is the resulting video showing the interface swapping back and forth from light to dark:


https://github.com/user-attachments/assets/d52da13c-9a39-48d9-8be7-ccfe37745b6f

## Issues fixed

n/a

markrickert commented:
Adde docs PR to document new feature - https://github.com/mobile-dev-inc/maestro-docs/pull/133

morganick commented:
@markrickert does this handle specifying the device?

markrickert commented:
> @markrickert does this handle specifying the device?

@morganick On iOS, it should pass through the device id and if you don't pass one, it'll use "booted" as a default. I'm unsure how android handles this. 

https://github.com/mobile-dev-inc/Maestro/pull/2507/files#diff-25ebfb9b3069b37481b09445ea8c4458fdebdc760b55ef6e9af537f0fc1e9221R518-R527

morganick commented:
> > @markrickert does this handle specifying the device?
> 
> @morganick On iOS, it should pass through the device id and if you don't pass one, it'll use "booted" as a default. I'm unsure how android handles this.
> 
> https://github.com/mobile-dev-inc/Maestro/pull/2507/files#diff-25ebfb9b3069b37481b09445ea8c4458fdebdc760b55ef6e9af537f0fc1e9221R518-R527

@markrickert Yeah, I saw the `deviceId` portion on iOS, but trying to run your commands when I have more than one android device results in `adb: more than one device/emulator`. It needs the `-s <serial-number>` for the specified device. I didn't see how this was handled on the android side either.

markrickert commented:
@morganick I believe i've addressed the problem. I was using the regular `shell` command when i should have been using `dadb.shell` to reference the currently active emulator instance.

https://github.com/mobile-dev-inc/Maestro/pull/2507/commits/fd264a598d4ae8587a73d7408f8d1dbae11d7b82

Leland-Takamine commented on maestro-orchestra/src/main/java/maestro/orchestra/yaml/YamlSetDarkMode.kt:
```diff
@@ -0,0 +1,72 @@
+            "setDarkMode command takes either: \n" +
+                    "\t1. enabled: To enable dark mode\n" +
+                    "\t2. disabled: To disable dark mode\n" +
+                    "\t3. value: To set dark mode to a specific value (enabled or disabled) \n" +
```
I was a bit confused about this line

markrickert replied:
I was copying the airplane mode function as it allows this syntax:

```yaml
- setDarkMode enabled
```

or

```yaml
- setDarkMode
    value: enabled
    label: "Something"
```

I'm definitely open to suggestions on how to improve this!

Leland-Takamine replied:
Oh yeah I think the syntax makes sense. The 3rd line of the error message is what was confusing me - seems redundant?

```
1. enabled: To enable dark mode
2. disabled: To disable dark mode
3. value: To set dark mode to a specific value (enabled or disabled)
```

Leland-Takamine replied:
Oh ok I think I understand now. Maybe it would be clearer to show those two styles of syntax as examples

Leland-Takamine commented:
> @morganick I believe i've addressed the problem. I was using the regular `shell` command when i should have been using `dadb.shell` to reference the currently active emulator instance.
> 
> [fd264a5](https://github.com/mobile-dev-inc/Maestro/commit/fd264a598d4ae8587a73d7408f8d1dbae11d7b82)

`shell` method was actually correct - it's just a helper that checks the exit code as well. You'd just pass in the adb shell command without the `adb shell` part since it's using Dadb.

```
    private fun shell(command: String): String {
        val response: AdbShellResponse = try {
            dadb.shell(command)
        } catch (e: IOException) {
            throw IOException(command, e)
        }

        if (response.exitCode != 0) {
            throw IOException("$command: ${response.allOutput}")
        }
        return response.output
    }
```

Leland-Takamine commented on maestro-client/src/main/java/maestro/drivers/IOSDriver.kt:
```diff
@@ -515,6 +515,16 @@ class IOSDriver(
 
+    override fun isDarkModeEnabled(): Boolean {
+        val deviceId = iosDevice.deviceId ?: "booted"
+        return XCRunnerCLIUtils.isDarkModeEnabled(deviceId)
```
Is there a way to do this via xcuitest? We're trying to move away from simctl commands since they won't work for physical devices

markrickert replied:
I'll look into this tomorrow to see if there's a way to do it. Not sure at the moment.

markrickert replied:
@Leland-Takamine I found this reference already inside the repo's source code, but i have no idea how i would go about using it 🤷  https://github.com/mobile-dev-inc/Maestro/blob/main/maestro-ios-xctest-runner/maestro-driver-iosUITests/PrivateHeaders/XCTest/XCUIDevice.h#L18-L22

markrickert replied:
I've updated the PR to utilize the XCTest runner framework instead of simctl for ios! 🎉

Fishbowler commented on maestro-orchestra/src/test/resources/YamlCommandReaderTest/023_labels.yaml:
```diff
@@ -124,6 +124,11 @@ appId: com.example.app
+      value: enabled
+      label: "Turn on dark mode for testing"
+- toggleDarkMode:
+      label: "Toggle dark mode for testing"
```
Matches what we do for Airplane mode. I like the consistency. But I've never used the `toggle` variant.

markrickert replied:
Yeah, left it in for consistency with airplane mode.

Fishbowler commented on maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt:
```diff
@@ -1026,6 +1026,39 @@ data class ToggleAirplaneModeCommand(
+        }
+
+    override fun evaluateScripts(jsEngine: JsEngine): Command {
+        return this
```
I wanna set _everything_ via JS, because at some point, someone _will need this_.

markrickert committed:
feat(ios): add XCUITest appearance endpoint for setDarkMode

Adds a setAppearance/appearance route pair to the XCUITest runner,
backed by XCUIDevice.shared.appearance (iOS 15+). Replaces the
simctl-based approach originally proposed.

markrickert committed:
feat(ios): wire dark mode through the Kotlin XCUITest client

XCTestDriverClient gets setAppearance()/getAppearance() calling the new runner endpoints; IOSDevice interface gains
isDarkModeEnabled()/setAppearance(), implemented in XCTestIOSDevice and delegated through LocalIOSDevice. Stubs added to SimctlIOSDevice and the experimental DeviceControlIOSDevice for interface compliance.

markrickert committed:
feat: add setDarkMode/isDarkModeEnabled to the Driver interface

Android reads/writes via `cmd uimode night`. iOS delegates to the new XCUITest appearance endpoint. Web/CdpWeb are no-ops.

markrickert committed:
feat: add setDarkMode/toggleDarkMode commands

SetDarkModeCommand/ToggleDarkModeCommand mirror the existing airplane-mode command pair end to end (Commands, MaestroCommand, Orchestra dispatch, YAML parsing).

YamlSetDarkMode's invalid-input error message shows the two accepted YAML syntaxes directly instead of the redundant three-bullet list airplane mode uses.

markrickert committed:
test: cover setDarkMode/toggleDarkMode

YamlCommandReaderTest + 023_labels.yaml exercise labeled YAML parsing. IntegrationTest Case 152 (121 was already taken on current main) runs the full enable/disable/toggle sequence against FakeDriver and asserts the resulting isDarkModeEnabled() state, rather than only assertNoInteraction() as the original #2507 test did.

markrickert committed:
feat(web): implement setDarkMode/isDarkModeEnabled via CDP media emulation

Was a no-op stub. CdpWebDriver already has the Selenium DevTools Emulation plumbing wired up for setLocation (geolocation override), so this follows the same pattern: Emulation.setEmulatedMedia forces the prefers-color-scheme media feature, and isDarkModeEnabled reads it back via the existing executeJS helper (window.matchMedia).

markrickert committed:
feat: add assertDarkMode and assertLightMode commands

markrickert committed:
refactor: merge assertDarkMode/assertLightMode into one helper

assertDarkModeCommand() and assertLightModeCommand() were copy-pasted with only the condition negated and message text swapped.

markrickert commented on maestro-test/src/test/kotlin/maestro/test/IntegrationTest.kt:
```diff
@@ -3240,6 +3244,20 @@ class IntegrationTest {
+
+        // Then
+        // No test failures
+        driver.assertNoInteraction()
```
yup! i've got the assertion commands all wired up now!

pedro18x reviewed (approved):
This is looking great, Mark, and thanks for seeing it through after Chain React. I went through the reworked commits end to end: the command plumbing, YAML models, Orchestra dispatch, and the assertion logic (real device read-back with a clear failure and debug message) all read well, and the integration cases cover the behavior. The iOS 15 precondition guard surfaces a clear "requires iOS 15+" message rather than an opaque failure, and keeping `setDarkMode` literal-only mirrors `setAirplaneMode` nicely for consistency.

One optional follow-up, definitely not a blocker: the web `CdpWebDriver` dark-mode path (`setDarkMode` / `isDarkModeEnabled`) isn't exercised by the integration cases since they run on `FakeDriver`. That's consistent with how the rest of `CdpWebDriver` is tested, so no objection to merging as-is, but a small test hitting the real prefers-color-scheme emulation and the `matchMedia` read-back would be a nice addition next time you're in here.

Approving. Nice work.
